backend/src/oauth/AuthorizationCodeModelImpl.ts

```diff
@@ -0,0 +1,394 @@
+      grants: clientFromDB.grants.split(',').map(grant => grant.trim()),
+    };
+  }
+
+  async generateAccessToken(client: Client, user: User, scope: string | string[]): Promise<string> {
+    const nowTs = Math.floor(Date.now() / 1000);
+    const expiresTs = nowTs + this.config.accessTokenTtlSeconds;
+    return AuthorizationCodeModelImpl.generateJwtToken(
+      user.id.toString(),
+      expiresTs,
+      nowTs,
+      client.id,
+      scope,
+      TokenType.Access,
+      {
+        client,
+        user
+      }
+    );
+  }
+
+  async getAccessToken(accessToken: string, callback?: Callback<Token>): Promise<Falsey | Token> {
+    try {
+      const {
+        aud,
+        exp,
+        iat,
+        sub,
+        scope,
+        type
+      } = jwt.verify(accessToken, process.env.JWT_SECRET_KEY) as JwtPayload;
+
+      const clientId = aud.toString();
+      const userId = parseInt(sub, 10);
+
+      // Verify token validity (returns Error object on error and grants on success)
+      const result = await AuthorizationCodeModelImpl.verifyTokenValidity(
+          this.repo,
+          type,
+          TokenType.Access,
+          exp,
+          iat,
+          clientId,
+          userId
+      );
+      if (result instanceof Error) {
+        throw result;
+      }
+
+      const decoded: Token = {
+        accessToken,
+        accessTokenExpiresAt: new Date(exp * 1000),
+        scope,
+        client: {
+          id: clientId,
+          grants: result ? result : []
+        },
+        user: { id: userId }
+      };
+      if (callback) {
+        callback(null, decoded);
+      }
+      return decoded;
+    } catch (error) {
+      if (callback) {
+        callback(error, null);
+      }
+    }
+  }
+
+  async saveToken(token: Token, client: Client, user: User): Promise<Token | Falsey> {
+    return {
+      accessToken: token.accessToken,
+      accessTokenExpiresAt: token.accessTokenExpiresAt,
+      refreshToken: token.refreshToken,
+      client: { id: client.id, grants: client.grants ? client.grants : [] },
+      user: { id: user.id },
+      scope: token.scope,
+    };
+  }
+
+  async verifyScope(token: Token, scope: string | string[], callback?: Callback<boolean>): Promise<boolean> {
+    // Convert the token's scopes into an array.
+    let tokenScopes: string[] = [];
+    if (token && token.scope) {
+      if (typeof token.scope === 'string') {
+        tokenScopes = token.scope.split(' ').filter((s) => s.trim());
+      } else if (Array.isArray(token.scope)) {
+        tokenScopes = token.scope;
+      }
+    }
+
+    // Convert the required scope(s) into an array.
+    const requiredScopes = typeof scope === 'string'
+      ? scope.split(' ').filter((s) => s.trim())
+      : scope;
+
+    const valid = requiredScopes.every(requiredScope => {
+      return tokenScopes.some(tokenScope => {
+        // Exact match check.
+        if (tokenScope === requiredScope) {
+          return true;
```

Comment:
IMPORTANT! callback must be called, if provided, otherwise the request will hang